misk-feature-testing/src/main/kotlin/misk/feature/testing/FakeFeatureFlags.kt

```diff
@@ -77,76 +77,171 @@ class FakeFeatureFlags @Inject constructor(private val moshi : Provider<Moshi>)
   override fun <T : Enum<T>> getEnum(feature: Feature, clazz: Class<T>): T = getEnum(feature, KEY, clazz)
   override fun <T> getJson(feature: Feature, clazz: Class<T>): T = getJson(feature, KEY, clazz)
 
-  private fun get(feature: Feature, key: String): Any? {
+  private fun <V> getOrDefault(
+    feature: Feature,
+    key: String,
+    defaultValue: V,
+    attributes: Attributes = defaultAttributes
+  ): V {
     FeatureFlagValidation.checkValidKey(feature, key)
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides[MapKey(feature)]
-    }
+    @Suppress("unchecked_cast")
+    return (get(feature, key, attributes) ?: defaultValue) as V
   }
 
-  private fun <V> getOrDefault(feature: Feature, key: String, defaultValue: V): V {
+  private fun get(
+    feature: Feature,
+    key: String,
+    attributes: Attributes = defaultAttributes
+  ): Any? {
     FeatureFlagValidation.checkValidKey(feature, key)
-    @Suppress("unchecked_cast")
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides.getOrDefault(MapKey(feature), defaultValue)
-    } as V
+    // Override value lookup is structured in 3 levels: feature, key, and attributes level. The
+    // logic to get the override value is as follows:
+    //
+    // 1. Return feature level override if there is no override at the key level.
+    // 2. if there is an override at the key level, then:
+    //   2.1 Return the override value associated with the attribute override that contains
+    //       all the attributes of the given attributes.
+    //   2.2 If there is no match, return the override value defined at the key level.
+    val overrideMapValues = overrides[MapKey(feature, key)]
+        ?: return overrides[MapKey(feature)]?.let { it.first().value }
+
+    val overrideMapValuesCopy = PriorityQueue(overrideMapValues)
+    var currentMapValue = overrideMapValuesCopy.poll()
+    while (overrideMapValuesCopy.isNotEmpty() || currentMapValue != null) {
+      if (attributes.text.entries.containsAll(currentMapValue.attributes.text.entries)) { break }
+      currentMapValue = overrideMapValuesCopy.poll()
+    }
+    return currentMapValue.value
   }
 
-  fun override(feature: Feature, value: Boolean) {
-    overrides[MapKey(feature)] = value
-  }
+  fun override(
+    feature: Feature,
+    value: Boolean
+  ) = override<Boolean>(feature, value)
```

Comment:
Is this a behavior change that causes all overrides to be stored as JSON?

Reply:
Previously they weren’t!

Reply:
nope, the execution flow will be:

1. `subject.override(Feature(name="foo"), true)`
1. `FakeFeatureFlags#override(feature: Feature, value: Boolean)` method Line 121
1. `FakeFeatureFlags#<T>override(feature: Feature, value: T)` method Line 141
1. `FakeFeatureFlags#<T>overrideKey(feature: Feature, key: String, value: T, attributes: Attributes = defaultAttributes)` method Line 204

One thing I could do to simplify the implementation is to get rid off `3.` and call `4.` directly in the implementation of methods `2.`